app-marketplace-v2-spec/structs.go

```diff
@@ -0,0 +1,125 @@
+package cloudappmanifest
+
+type App struct {
+	DisplayName string
+	Description string
+	SiteURL     string
+
+	CommandHandlers []Function
+
+	SlashCommands []SlashCommand
+	ModalDialogs  []ModalDialog
+	BotDialogs    []BotDialog
+	Widgets       []Widget
+
```

Comment:
I am starting to like having separated the functions from the widgets. We call define the functions by an internal name, and the endpoint they should reach. Then a widget is defined by type/location (whichever name we finally use), callable type (whether it runs a function or opens a dialog), and the name of the callable. The arguments passed will be the relevant on the widget context. Going down to manifest definition...

```
type Function struct {
    Name string  // e.g. create_meeting
    URL string  // e.g. https://myservice.com/start
    // Maybe extra static args we want to pass either way? Not sure how useful
    Args []Arg // e.g. {source: mattermost}
}

type Widget struct {
    Location string // e.g. CHANNEL_HEADER
    CallableType string // e.g. FUNCTION
    Callable string // e.g. create_meeting
    Extra interface{} // location dependant extra information (icon, text...)
}
```

Client will not receive the functions, and when a widget is activated, it just tells the server "run command X from app Y with these arguments".

Reply:
`Name` is the URL (with some common prefix). `CallableType` - maybe, I suppose would need it to decode even in go.

Reply:
My point of giving them names is to make it more readable (and writable) the manifest.
e.g., I have registered the function create meeting, which is called by the channel header widget, and the start slash command.

The way I see it, the alternative would be something like:
I have registered the function that triggers the url `http://myservice.com/start`, and a widget that calls the function `http://myservice.com/start` and a slash command that calls the function `http://myservice.com/start`.